msal/wam.py

```diff
@@ -0,0 +1,189 @@
+"""This module acquires token via WAM, with the help of Mid-tier.
+
+Mid-tier supports these Windows versions
+https://github.com/AzureAD/microsoft-authentication-library-for-cpp/pull/2406/files
+"""
+from threading import Event
+import json
+import logging
+
+import pymsalruntime  # ImportError would be raised on unsupported platforms such as Windows 8
+    # Its API description is available in site-packages/pymsalruntime/PyMsalRuntime.pyi
+
+logger = logging.getLogger(__name__)
+
+
+class RedirectUriError(ValueError):
+    pass
+
+
+class _CallbackData:
+    def __init__(self):
+        self.signal = Event()
+        self.auth_result = None
+
+    def complete(self, auth_result):
+        self.signal.set()
+        self.auth_result = auth_result
+
+
+def _convert_error(error, client_id):
+    context = error.get_context()  # Available since pymsalruntime 0.0.4
+    if "AADSTS50011" in context:  # In WAM, this could happen on both interactive and silent flows
```

Comment:
FYI, this is error message that server should return based on their test: https://msazure.visualstudio.com/DefaultCollection/One/_git/226286d5-a05d-446c-a950-4c323c15e588?path=/src/Test/BvtFvt.Microsoft.AzureAD.Sts/OAuth2/OAuth2EndToEndTests.cs&version=GBmaster&line=3629&lineEnd=3632&lineStartColumn=41&lineEndColumn=113&lineStyle=plain&_a=contents